resolver-dns/src/main/java/io/netty/resolver/dns/DnsCache.java

```diff
@@ -49,17 +49,19 @@
      * @param additionals the additional records
      * @return the cached entries
      */
-    List<DnsCacheEntry> get(String hostname, DnsRecord[] additionals);
+    List<? extends DnsCacheEntry> get(String hostname, DnsRecord[] additionals);
 
     /**
-     * Cache a resolved address for a given hostname.
+     * Create a new {@link DnsCacheEntry} and cache a resolved address for a given hostname.
      * @param hostname the hostname
      * @param additionals the additional records
      * @param address the resolved address
      * @param originalTtl the TLL as returned by the DNS server
      * @param loop the {@link EventLoop} used to register the TTL timeout
+     * @return The {@link DnsCacheEntry} corresponding to this cache entry.
      */
-    void cache(String hostname, DnsRecord[] additionals, InetAddress address, long originalTtl, EventLoop loop);
+    DnsCacheEntry cache(String hostname, DnsRecord[] additionals, InetAddress address, long originalTtl,
```

Comment:
I debated doing this but it isn't currently necessary. The primary user of the cache (Netty) does not need the cache entry for the failure case. This allows the cache implementation to not even allocate an object if not necessary. We could also allow `null` to be returned but I figured it would be less complicated to just keep `void` until it is actually needed.

I'm not strong here and I can change if you prefer.